core/src/main/java/org/elasticsearch/index/percolator/PercolatorFieldMapper.java

```diff
@@ -60,19 +73,28 @@ public Builder(QueryShardContext queryShardContext) {
         @Override
         public PercolatorFieldMapper build(BuilderContext context) {
             context.path().add(name);
-            KeywordFieldMapper extractedTermsField = createStringFieldBuilder(EXTRACTED_TERMS_FIELD_NAME).build(context);
-            KeywordFieldMapper unknownQueryField = createStringFieldBuilder(UNKNOWN_QUERY_FIELD_NAME).build(context);
+            KeywordFieldMapper extractedTermsField = createExtractQueryFieldBuilder(EXTRACTED_TERMS_FIELD_NAME).build(context);
+            KeywordFieldMapper unknownQueryField = createExtractQueryFieldBuilder(UNKNOWN_QUERY_FIELD_NAME).build(context);
+            KeywordFieldMapper queryBuilderField = createQueryBuilderFieldBuilder().build(context);
             context.path().remove();
-            return new PercolatorFieldMapper(name(), fieldType, defaultFieldType, context.indexSettings(), multiFieldsBuilder.build(this, context), copyTo, queryShardContext, extractedTermsField, unknownQueryField);
+            return new PercolatorFieldMapper(name(), fieldType, defaultFieldType, context.indexSettings(), multiFieldsBuilder.build(this, context), copyTo, queryShardContext, extractedTermsField, unknownQueryField, queryBuilderField);
         }
 
-        static KeywordFieldMapper.Builder createStringFieldBuilder(String name) {
+        static KeywordFieldMapper.Builder createExtractQueryFieldBuilder(String name) {
             KeywordFieldMapper.Builder queryMetaDataFieldBuilder = new KeywordFieldMapper.Builder(name);
             queryMetaDataFieldBuilder.docValues(false);
             queryMetaDataFieldBuilder.store(false);
             queryMetaDataFieldBuilder.indexOptions(IndexOptions.DOCS);
             return queryMetaDataFieldBuilder;
         }
+
+        static KeywordFieldMapper.Builder createQueryBuilderFieldBuilder() {
+            KeywordFieldMapper.Builder queryMetaDataFieldBuilder = new KeywordFieldMapper.Builder(QUERY_BUILDER_FIELD_NAME);
+            queryMetaDataFieldBuilder.docValues(false);
+            queryMetaDataFieldBuilder.store(true);
```

Comment:
In the PercolatorFieldMapper we now store the QueryBuilder itself (the byte[] representation of its toXContent), so that at index warm time in PercolatorQueryCache we just need to load this and not the entire source which we would then need to parse first to extract the `query` field object.